# Ames Housing Regression Project

## Test 1: Local Data Ingestion and Structural Verification
In this test, I intend to load the raw dataset from my local directory into a Pandas DataFrame. My objective is to verify that the file path is correct, ensure the data is loaded without corruption, and inspect the initial shape (total rows and columns) of the dataset.

In [14]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Modelos e Métricas para a Regressão
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression  # O modelo assertivo do seu projeto anterior!
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

print("[SUCCESS] Todas as bibliotecas foram carregadas com sucesso!")

[SUCCESS] Todas as bibliotecas foram carregadas com sucesso!


In [15]:
base_housing = pd.read_csv('../data/raw/train.csv')
base_housing

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1455,1456,60,RL,62.0,7917,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,8,2007,WD,Normal,175000
1456,1457,20,RL,85.0,13175,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,MnPrv,NaN,0,2,2010,WD,Normal,210000
1457,1458,70,RL,66.0,9042,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,GdPrv,Shed,2500,5,2010,WD,Normal,266500
1458,1459,20,RL,68.0,9717,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,4,2010,WD,Normal,142125


In [16]:
base_housing.info()

<class 'pandas.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   str    
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   str    
 6   Alley          91 non-null     str    
 7   LotShape       1460 non-null   str    
 8   LandContour    1460 non-null   str    
 9   Utilities      1460 non-null   str    
 10  LotConfig      1460 non-null   str    
 11  LandSlope      1460 non-null   str    
 12  Neighborhood   1460 non-null   str    
 13  Condition1     1460 non-null   str    
 14  Condition2     1460 non-null   str    
 15  BldgType       1460 non-null   str    
 16  HouseStyle     1460 non-null   str    
 17  OverallQual    1460 non-null   int64  
 18  OverallCond    1460

## 2. Exploratory Data Analysis (EDA)

### 2.1 Dataset Structure Evaluation
By running `base_housing.info()`, we verify that the dataset contains 1,460 records and 81 features (columns 0 to 80). 
* **Target Variable:** `SalePrice` (the home price we want to predict).
* **Feature Types:** A mix of continuous numerical variables (integers/floats) and categorical attributes (objects).
* **Missing Data:** Several columns show a `non-null` count lower than 1,460, indicating the presence of missing values (`NaN`) that require an imputation strategy.

In [17]:
base_housing.head()

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


In [18]:
base_housing.isnull().sum()

Id                 0
MSSubClass         0
MSZoning           0
LotFrontage      259
LotArea            0
                ... 
MoSold             0
YrSold             0
SaleType           0
SaleCondition      0
SalePrice          0
Length: 81, dtype: int64

### 2.2 Missing Values Identification
By analyzing `base_housing.isnull().sum()`, we can locate the exact volume of missing data per feature. 
* **Key Finding:** The feature `LotFrontage` shows a significant amount of missing data, with **259 null values** (`NaN`).
* **Next Step:** Since the standard output truncates the rows, we will filter and display only the columns that actually contain missing values to plan our imputation strategy.

In [19]:
# Filter and show only columns that contain missing values (NaN)
missing_counts= base_housing.isnull().sum()
missing_counts[missing_counts >0]

LotFrontage      259
Alley           1369
MasVnrType       872
MasVnrArea         8
BsmtQual          37
BsmtCond          37
BsmtExposure      38
BsmtFinType1      37
BsmtFinType2      38
Electrical         1
FireplaceQu      690
GarageType        81
GarageYrBlt       81
GarageFinish      81
GarageQual        81
GarageCond        81
PoolQC          1453
Fence           1179
MiscFeature     1406
dtype: int64

### 2.3 Comprehensive Missing Data Analysis
By filtering the features with missing values, we observe a clear distinction between random missingness and structural missingness:

* **Structural Missingness (Structural NaNs):** Features like `PoolQC` (1,453), `MiscFeature` (1,406), `Alley` (1,369), `Fence` (1,179), and `FireplaceQu` (690) have high null counts because most houses simply do not possess these physical attributes (e.g., no pool, no fence). The same pattern applies to the 81 missing rows in the `Garage` variables and 37/38 rows in the `Basement` variables.
* **Accidental/Random Missingness:** Features like `Electrical` (1 missing row), `MasVnrArea` (8 missing rows), and `LotFrontage` (259 missing rows) represent real missing information that must be handled via mathematical or statistical imputation.

## 3. Data Cleaning and Imputation Strategy
In this section, I will handle the missing values identified during the exploratory phase. My strategy is split into two parts:
1. **Categorical Imputation:** Replace `NaN` with a specific category (e.g., `'None'` or `'No'`) for houses that do not have the physical attribute.
2. **Numerical Imputation:** Replace `NaN` with statistical metrics (like the median) for continuous variables.

In [21]:
# Check the most frequent value(mode) in the Electrical column
most_frequent_electrical = base_housing['Electrical'].mode()[0]

In [22]:
# Fill the single NaN value with the most frequent value
base_housing['Electrical'] = base_housing['Electrical'].fillna(most_frequent_electrical)

In [23]:
# Verify if there are any missing values left in Electrical
base_housing['Electrical'].isnull().sum()

np.int64(0)

### 3.1 Imputation for Electrical Column
Since the `Electrical` feature had only 1 missing value, I identified the most frequent electrical system (the mode) in the dataset and used it to fill this single gap. After the imputation, I verified that the remaining null count for this column is 0.

In [26]:
# Inspect the unique values and counts for a garage column before imputation
base_housing['GarageType'].value_counts(dropna=False)

GarageType
Attchd     870
Detchd     387
BuiltIn     88
NaN         81
Basment     19
CarPort      9
2Types       6
Name: count, dtype: int64

In [27]:
# List of categorical garage columns to treat
garage_categorical_cols = ['GarageType', 'GarageFinish', 'GarageQual', 'GarageCond']

# Fill NaN with 'None' for houses with no garage
for col in garage_categorical_cols:
    base_housing[col] = base_housing[col].fillna('None')

In [28]:
# Verify if NaN changed to 'None'
base_housing['GarageType'].value_counts(dropna=False)

GarageType
Attchd     870
Detchd     387
BuiltIn     88
None        81
Basment     19
CarPort      9
2Types       6
Name: count, dtype: int64

### 3.2 Imputation for Categorical Garage Features
I identified 4 categorical garage features (`GarageType`, `GarageFinish`, `GarageQual`, and `GarageCond`) with 81 missing values. Since these missing values represent the absence of a garage, I filled the `NaN` entries with the string `'None'`. A final check confirmed that all `NaN` values in these columns were successfully replaced.

In [29]:
# Check de summary statistics for the garage year built column
base_housing['GarageYrBlt'].describe()

count    1379.000000
mean     1978.506164
std        24.689725
min      1900.000000
25%      1961.000000
50%      1980.000000
75%      2002.000000
max      2010.000000
Name: GarageYrBlt, dtype: float64